### PR TITLE
Create plugin script

### PR DESCRIPTION
**Pull request by oxarbitrage**
When working with plugins the developer find himself doing the same renaming work too many times. 

I found a script in eos that actually do the job of automating plugin creation: https://github.com/EOSIO/eos/blob/master/plugins/eosio-make_new_plugin.sh

So i copied over and added some further instructions to get up and running in bitshares after the files are all in place.

The template is a plugin with implementation class, most of our plugins are like that even if there are others that don't follow this pattern i think this is the way most of them can be built.

It haves some minimal functionality on how to add an argument and an onBlock function that will execute every block.

For further examples the developer will check other plugins, unfortunately we have no plugin that exposes api calls in the plugin itself. This is only visible on @xeroc hello plugin:

https://github.com/bitshares/bitshares-core/tree/hello_plugin/libraries/plugins

I think we should encourage the use of the api in the plugin itself rather than adding to `api.cpp` but i didn't added this to template as most plugins will not want it, tried to keep it simple.

Sample script output:

```
alfredo@alfredo-Inspiron-5559 ~/CLionProjects/template_plugin/libraries/plugins $ ./make_new_plugin.sh stoploss
Copying template...
Renaming files/directories...
Renaming in files...
Done! stoploss is ready.
Next steps:
1- Add 'add_subdirectory( stoploss )' to CmakeLists.txt in this directory.
2- Add 'graphene_stoploss' to ../../programs/witness_node/CMakeLists.txt with the other plugins.
3- Include plugin header file '#include <graphene/stoploss/stoploss.hpp>' to ../../programs/witness_node/main.cpp.
4- Initialize plugin with the others with 'auto stoploss_plug = node->register_plugin<stoploss::stoploss>();' in ../../programs/witness_node/main.cpp
5- cmake and make
5- Start plugin with './../programs/witness_node/witness_node --plugins "stoploss"'. After the seed nodes are added you start to see see a msgs from the plugin 'onBlock' 
alfredo@alfredo-Inspiron-5559 ~/CLionProjects/template_plugin/libraries/plugins $ 
```

### CODE REVIEW

**Comment by clockworkgr** [comment]
Nice!

**Comment by pmconrad** [comment]
It would be nice to have the script look up used database space ID from other plugins and automatically select a new one.

**Comment by oxarbitrage** [comment]
I am not very good at bash but created a function in script to detect biggest space_id and sum 1 to it. This number will be used in the template being created. This is at https://github.com/bitshares/bitshares-core/pull/1302/commits/64505a8b3b90a51ae3b4a24d8feb4ba5db0d9541

File: libraries/plugins/make_new_plugin.sh [hunk 1]
```diff
@@ -21,24 +21,24 @@ next_space_id () {
...
+    echo "... where my_new_plugin is the name of the plugin you want to create"
     exit 1
 fi
 
-pluginName=$1
+pluginName="$1"
 
-echo Copying template...
+echo "Copying template..."
 cp -r template_plugin $pluginName
```
Comment (pmconrad) [inline libraries/plugins/make_new_plugin.sh]: Please quote $pluginName here and in the following lines.

File: libraries/plugins/make_new_plugin.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+fi
+
+pluginName="$1"
+
+echo "Copying template..."
+cp -r template_plugin "$pluginName"
+
+echo "Renaming files/directories..."
+mv "$pluginName/include/graphene/template_plugin" "$pluginName/include/graphene/$pluginName"
+for file in `find "$pluginName" -type f -name '*template_plugin*'`; do mv $file `sed s/template_plugin/"$pluginName"/g <<< $file`; done;
```
Comment (pmconrad) [inline libraries/plugins/make_new_plugin.sh]: Please quote ``$file``

File: libraries/plugins/make_new_plugin.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+echo "Renaming in files..."
+find $pluginName -type f -exec sed -i "s/template_plugin/$pluginName/g" {} \;
```
Comment (pmconrad) [inline libraries/plugins/make_new_plugin.sh]: Please quote ``$pluginName``

File: libraries/plugins/make_new_plugin.sh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+echo "Assigning next available SPACE_ID..."
+next_space_id
+find $pluginName -type f -exec sed -i "s/@SPACE_ID@/$?/g" {} \;
```
Comment (pmconrad) [inline libraries/plugins/make_new_plugin.sh]: Please quote ``$pluginName``

**Comment by pmconrad** [comment]
Excellent, tested successfully. Some of the "next steps" could also be automated, but I think it might be better to leave these as manual steps.
Thanks!

**Comment by oxarbitrage** [comment]
thanks @pmconrad :)

**Pull request merged**